src/main/state/index.js

```diff
@@ -36,21 +26,61 @@ function onLoadState(key) {
   return decryptFromFile(key, storeFileEncryptionAlgorithm, storeFilePath);
 }
 
+/**
+ * Create a onSaveState handler.
+ * Used to keep track of pending save requests that must be fulfilled before exiting the app.
+ */
+function OnSaveState() {
+  this.pendingSave = null;
+  this.currentPromise = null;
+
+  /**
+   * Handler function for the SAVE_STATE event.
+   *
+   * @async
+   * @param {Object} state - State tree.
+   */
+  this.onSaveState = async (state) => {
+    if (this.currentPromise) {
+      this.pendingSave = state;
+      return this.currentPromise;
+    }
+
+    this.currentPromise = encryptToFile(state, KEY, storeFileEncryptionAlgorithm, storeFilePath)
+      .then(() => {
+        this.currentPromise = null;
+
+        if (this.pendingSave) {
+          const tree = this.pendingSave;
+          this.pendingSave = null;
+          return this.onSaveState(tree);
+        }
+
+        return null;
+      });
+
+    return this.currentPromise;
```

Comment:
arf this means I will have to make an event to get the storefilepath lol

Reply:
Why ? I don't know why you need it renderer-side.